src/pyproject_hooks/_impl.py

```diff
@@ -6,31 +6,37 @@
 from os.path import abspath
 from os.path import join as pjoin
 from subprocess import STDOUT, check_call, check_output
+from typing import Any, Dict, Iterator, List, Optional, Protocol
 
 from ._in_process import _in_proc_script_path
 
 
-def write_json(obj, path, **kwargs):
+def write_json(obj: Dict[str, Any], path: str, **kwargs) -> None:
     with open(path, "w", encoding="utf-8") as f:
         json.dump(obj, f, **kwargs)
 
 
-def read_json(path):
+def read_json(path: str) -> Dict[str, Any]:
     with open(path, encoding="utf-8") as f:
         return json.load(f)
 
 
 class BackendUnavailable(Exception):
     """Will be raised if the backend cannot be imported in the hook process."""
 
-    def __init__(self, traceback):
+    def __init__(self, traceback: str) -> None:
         self.traceback = traceback
```

Comment:
Public attributes need to be annotated:

```suggestion
        self.traceback: str = traceback
```

Reply:
This shouldn’t be needed, type checkers infer this from the arguments if it is `__init__`.